### ISSUE TITLE
Inbox for non-LV actors?

### ISSUE THREAD

**Issue by DavidStaab** [featherweight/ftw-kernel#6]
`ftw_framework.h` exposes functions that interact with LabVIEW's event system to create the Inbox feature...what about actors in a mixed-language system? How is a Featherweight actor implemented in C, JS, Python, or any other viable language with a nanomsg binding? (Or just a websocket binding, for that matter)? Doesn't the Featherweight architecture depend critically on every actor having an Inbox?

**Comment by wirebirdlabs** [comment]
Quick honest question: is it clear that this library, ftw-kernel, is currently almost exclusively only useful when called from a LabVIEW CLFN?

Asked another way -- have you been considering wrapping this library, in say, Python's FFI, or calling from C?

Understanding what you're thinking here will help me answer this question better, but I think you may have discovered a bug this library misleading one into believing it is currently able to be wrapped in any number of language FFI's capable of C calling convention ...

**Comment by DavidStaab** [comment]
Yes, I'm interested in FTW as a framework for getting an LV actor to communicate with {another LV actor, a non-LV actor or service}. I don't see the value in yet-another-LV-to-LV-messaging-framework, especially not if it has to jump through several design hoops to circumvent bugs or simple limitations in LabVIEW. National Instruments has already made it relatively easy to get LV actors to talk to other LV actors on any topology, and the gaps in their platform have already been filled in by a dozen all-G messaging and actor frameworks.

**Comment by wirebirdlabs** [comment]
> Yes, I'm interested in FTW as a framework for getting an LV actor to communicate with {another LV actor, a non-LV actor or service}

Of course; that's what it does, but just not by using this library. This library is what equips LV with the capability to talk with other languages. The other languages don't need this library, because they already have good facilities for interoperability.

**Comment by DavidStaab** [comment]
FTW Actors expect to be contacted with their Inbox, don't they? There's no Inbox for a non-LV actor...so we have to create out-of-band messaging (from the perspective of the framework) in order to send messages to a non-LV actor?

**Comment by wirebirdlabs** [comment]
> There's no Inbox for a non-LV actor...so we have to create out-of-band messaging (from the perspective of the framework) in order to send messages to a non-LV actor?

This is simple; perhaps it's just unclear still.

Pseudocode for non-LV entity requesting an echo from `ftw-master\tests\core\FTW-TestSupport-Actor-EchoServer.vi`:

1. Open REP socket
2. Connect to Actor Instance address
3. nn_send( '{"FTW-Request-Header":"FTW-ActorStatus"}')
4. nn_recv() -> "{"FTW-Request-Header":"FTW-ActorStatus", "FTW-Reply-Timestamp":"...", "FTW-Reply-RelativeTimer": .... }"

**Comment by DavidStaab** [comment]
Yes, it's easy. But it breaks the abstraction that a messaging framework normally offers: Go through the framework, and you don't have to worry about the specifics of transport, format, or how the entity on the other side was built. With ftw-kernel, I do have to worry about that: If the peer is a LV actor, it naturally wants me to talk to its Inbox. If it's a non-LV actor, it doesn't have an Inbox, so I'll have to know more about its implementation to even establish cursory information exchange with it. There isn't even a means of asking what interface it wants to use, because I can't talk to it until I know what interface it wants to use.
